address/address.go

```diff
@@ -47,327 +77,236 @@ const (
-	ErrInvalidBytes = errors.New("invalid bytes")
-)
+// MainnetPrefix is the main network prefix.
+var MainnetPrefix = "f"
 
-// NetworkFromString tries to convert the string representation of a network to
-// its binary representation.
-func NetworkFromString(input string) (Network, error) {
-	switch input {
-	case "fc":
-		return Mainnet, nil
-	case "tf":
-		return Testnet, nil
-	default:
-		return 0, ErrUnknownNetwork
-	}
-}
+// TestnetPrefix is the main network prefix.
+var TestnetPrefix = "t"
 
-// NetworkToString creates a human readable representation of the network.
-func NetworkToString(n Network) string {
-	switch n {
-	case Mainnet:
-		return "fc"
-	case Testnet:
-		return "tf"
-	default:
-		panic("invalid network")
-	}
-}
+// Protocol represents which protocol an address uses.
+type Protocol = byte
 
-// Address is the go type that represents an address in the filecoin network.
-type Address [Length]byte
+const (
+	// ID represents the address ID protocol.
+	ID Protocol = iota
+	// SECP256K1 represents the address SECP256K1 protocol.
+	SECP256K1
+	// Actor represents the address Actor protocol.
+	Actor
+	// BLS represents the address BLS protocol.
+	BLS
+)
 
-// NewMainnet constructs a new mainnet address.
-func NewMainnet(hash []byte) Address {
-	return New(Mainnet, hash)
+// Protocol returns the protocol used by the address.
+func (a Address) Protocol() Protocol {
+	return a.str[0]
 }
 
-// NewTestnet constructs a new testnet address.
-func NewTestnet(hash []byte) Address {
-	return New(Testnet, hash)
+// Payload returns the payload of the address.
+func (a Address) Payload() []byte {
+	return []byte(a.str[1:])
 }
 
-// New constructs a new address for the given nework.
-func New(network Network, hash []byte) Address {
-	var addr [Length]byte
-	addr[0] = network
-	addr[1] = Version
-	copy(addr[2:], hash)
-	return addr
+// Bytes returns the address as bytes.
+func (a Address) Bytes() []byte {
+	return []byte(a.str)
 }
 
-// NewFromString tries to parse a given string into a filecoin address.
-func NewFromString(s string) (Address, error) {
-	networkString, version, hash, err := decode(s)
-	if err != nil {
-		return Address{}, err
-	}
-
-	network, err := NetworkFromString(networkString)
+// String returns an address encoded as a string.
+func (a Address) String() string {
+	str, err := encode(Mainnet, a)
 	if err != nil {
-		return Address{}, err
-	}
-
-	if version != Version {
-		return Address{}, ErrUnknownVersion
+		panic(err)
 	}
-
-	return New(network, hash), nil
+	return str
 }
 
-// NewFromBytes tries to create an address from the given bytes.
-func NewFromBytes(raw []byte) (Address, error) {
-	if len(raw) != 22 {
-		return Address{}, ErrInvalidBytes
-	}
-
-	network := raw[0]
-	if network > Testnet {
-		return Address{}, ErrUnknownNetwork
-	}
+// Equal returns true if address `b` is equal to address.
+func (a Address) Equal(b Address) bool {
+	return bytes.Equal(a.Bytes(), b.Bytes())
```

Comment:
Elsewhere @whyrusleeping suggested this is unnecessary and we can use `==` (which should be documented)

Reply:
If addresses were fixed length arrays this would be true, but since they are variable length slices I believe `bytes.Equal` is required.